util/patch/patch.go

```diff
@@ -86,49 +83,167 @@ func (h *Helper) Patch(ctx context.Context, resource runtime.Object) error {
-		// to avoid uneccsary patching.
-		unstructured.RemoveNestedField(after, "status")
+	after, err := extractStatusWithoutConditions(h.after)
+	if err != nil {
+		return err
+	}
+	if !shouldPatch(before, after) {
+		return nil
 	}
+	return h.client.Status().Patch(ctx, resource.DeepCopyObject(), h.statusPatch)
+}
 
-	var errs []error
+// patchStatusConditions issues a patch if there is any changes to the conditions slice under
+// the status subresource. This is a special case and it's handled separately given that
+// we allow different controllers to act on conditions of the same object.
+//
+// This method has an internal backoff loop. When a conflict is detected, the method
+// asks the Client for the a new version of the resource we're trying to patch.
+//
+// Condition changes are then applied to the latest version of the object, and if there are
+// no unresolvable conflicts, the patch is sent again.
+func (h *Helper) patchStatusConditions(ctx context.Context, resource runtime.Object) error {
+	// Store the patch from the before/after object.
+	conditionPatch := conditions.NewPatch(
+		conditions.UnstructuredGetter(h.before),
+		conditions.UnstructuredGetter(h.after),
+	)
+
+	// Return early if there are no changes.
+	if conditionPatch.IsZero() {
+		return nil
+	}
 
-	if !reflect.DeepEqual(h.before, after) {
-		// only issue a Patch if the before and after resources (minus status) differ
-		if err := h.client.Patch(ctx, resource.DeepCopyObject(), h.resourcePatch); err != nil {
-			errs = append(errs, err)
-		}
+	// Make a copy of the object and store the key used if we have conflicts.
+	resourceCopy := resource.DeepCopyObject()
+	key, err := client.ObjectKeyFromObject(resourceCopy)
+	if err != nil {
+		return err
 	}
 
-	if (h.hasStatus || hasStatus) && !reflect.DeepEqual(h.beforeStatus, afterStatus) {
-		// only issue a Status Patch if the resource has a status and the beforeStatus
-		// and afterStatus copies differ
-		if err := h.client.Status().Patch(ctx, resource.DeepCopyObject(), h.statusPatch); err != nil {
-			errs = append(errs, err)
+	// Define and start a backoff loop to handle conflicts
+	// between controllers working on the same object.
+	//
+	// This has been copied from https://github.com/kubernetes/kubernetes/blob/release-1.16/pkg/controller/controller_utils.go#L86-L88.
+	backoff := wait.Backoff{
+		Steps:    5,
+		Duration: 100 * time.Millisecond,
+		Jitter:   1.0,
+	}
+	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		// Issue a patch.
+		err := h.client.Status().Patch(ctx, resourceCopy, h.conditionsPatch)
+		switch {
+		// Handle conflicts.
+		case apierrors.IsConflict(err):
+			// Only perform conflict resolution on objects that satisfy to the conditions.Setter interface.
+			conditionSetter, ok := resourceCopy.(conditions.Setter)
+			if !ok {
+				return false, errors.Wrapf(err, "cannot perform conflict resolution on unknown objects")
```

Comment:
```suggestion
				return false, errors.Wrapf(err, "cannot perform conflict resolution on objects not implementig the conditions.Setter interface")
```

Reply:
Well, that's true we look for the interface, although when using `corev1.Node` for example, we don't want to give the impression to the user that they can use our interface and get automatic conflict resolution

Reply:
fixed, ptal